File: JS/twoNumSum/twoNumSum1.js
// O(n) time | O(n) space
function twoNumSum(array, targetSum) {
  // save previously seen nums
  let prevNums = {};
	
	// iterate through the input array
	for (let num of array) {
    let deficit = targetSum - num;
    
    if (deficit in prevNums) {
      return [deficit, num];
    } else {
      prevNums[num] = num;
    }
  }
  
  // return empty array if no pair is found
  return [];
}

// Testing
let array = [1, 2, 3, 4, 5, 6, 7, 8, 9];
let targetSum = 17;

console.log(twoNumSum(array, targetSum));

File: Python/two_num_sum/two_num_sum1.py
# O(n^2) time | O(1) space
def two_num_sum(array, targetSum):
  for i in range(len(array) - 1):
    first_num = array[i]
    
    for j in range(i + 1, len(array)):
      second_num = array[j]
      if targetSum - first_num == second_num:
        return [first_num, second_num]
			
  # return an empty array if no pair is found
  return []

# Testing
ex = [3, 5, -4, 8, 11, 1, -1, 6]
print(two_num_sum(ex, 10))
